crates/ty_python_semantic/src/types/class.rs

```diff
@@ -2996,6 +3047,147 @@ impl<'db> From<ClassLiteral<'db>> for ClassType<'db> {
+        let new_name: &Name = &"__new__".into();
+        let lt_name: Name = "__lt__".into();
+        let le_name: Name = "__le__".into();
+        let gt_name: Name = "__gt__".into();
+        let ge_name: Name = "__ge__".into();
+        let eq_name: Name = "__eq__".into();
+        let repr_name: Name = "__repr__".into();
+        let hash_name: Name = "__hash__".into();
+
+        let order_methods = [lt_name, le_name, gt_name, ge_name];
+        let eq_methods = [eq_name];
+        let repr_methods = [repr_name];
+        let unsafe_hash_methods = [hash_name];
+
+        let use_def_map = index.use_def_map(class_body_scope.file_scope_id(db));
+        let table = place_table(db, class_body_scope);
+        let attribute_places_and_qualifiers =
+            use_def_map
+                .all_end_of_scope_symbol_declarations()
+                .map(|(symbol_id, declarations)| {
+                    let place_and_qual =
+                        place_from_declarations(db, declarations).ignore_conflicting_declarations();
+                    (symbol_id, place_and_qual)
+                })
+                .chain(use_def_map.all_end_of_scope_symbol_bindings().map(
+                    |(symbol_id, bindings)| (symbol_id, place_from_bindings(db, bindings).into()),
+                ))
+                .filter_map(|(symbol_id, place_and_qual)| {
+                    // TODO: Consider whether we need to worry about the Member variant here
+                    table
+                        .place(symbol_id)
+                        .as_symbol()
+                        .map(Symbol::name)
+                        .is_some_and(|name| ![init_name, new_name].contains(&name))
+                        .then_some(place_and_qual)
+                });
+
+        let extra_dataclass_members = [
+            (DataclassParams::ORDER, &order_methods[..]),
+            (DataclassParams::EQ, &eq_methods[..]),
+            (DataclassParams::REPR, &repr_methods[..]),
+            (DataclassParams::UNSAFE_HASH, &unsafe_hash_methods[..]),
+        ]
+        .into_iter()
+        .flat_map(|(param, members)| {
+            self.dataclass_params(db)
+                .is_some_and(|params| params.contains(param))
+                .then_some(members)
+                .into_iter()
+                .flatten()
+        });
+
+        let attribute_names = attribute_scopes(db, self.body_scope(db))
+            .flat_map(|function_scope_id| {
+                index
+                    .place_table(function_scope_id)
+                    .members()
+                    .filter_map(|member| member.as_instance_attribute())
+                    .filter(|name| *name != init_name && *name != new_name)
+                    .map(std::string::ToString::to_string)
+                    .collect::<Vec<_>>()
+            })
+            .dedup();
+
+        let attribute_variances = attribute_names
+            .chain(extra_dataclass_members.map(std::string::ToString::to_string))
```

Comment:
No tests fail if I remove this line. I don't think there is actually any way for the signatures of synthesized ordering methods, `__eq__`, `__repr__`, or `__hash__` to ever contain a typevar, so I don't think we can write a test that would require this line. Given that, I'm inclined to remove this extra logic and replace it with a comment that dataclasses may have some additional synthesized methods which we don't consider since they won't affect variance.